src/main/java/net/neoforged/neoforge/registries/RegistryManager.java

```diff
@@ -233,4 +270,68 @@ public enum SnapshotType {
          */
         FULL
     }
+
+    @ApiStatus.Internal
+    public static <R> void handleDataMapSync(final RegistryDataMapSyncPayload<R> payload, final PlayPayloadContext context) {
+        context.workHandler().submitAsync(() -> {
+            final BaseMappedRegistry<R> registry = (BaseMappedRegistry<R>) context.level().orElseThrow().registryAccess()
+                    .registryOrThrow(payload.registryKey());
+            registry.dataMaps.clear();
+            payload.dataMaps().forEach((attachKey, maps) -> registry.dataMaps.put(getDataMap(payload.registryKey(), attachKey), Collections.unmodifiableMap(maps)));
+        }).exceptionally(ex -> {
+            context.packetHandler().disconnect(Component.translatable("neoforge.network.data_maps.failed", payload.registryKey().location(), ex.getMessage()));
+            LOGGER.error("Failed to handle registry data map sync: ", ex);
+            return null;
+        });
+    }
+
+    @ApiStatus.Internal
+    public static void handleKnownDataMaps(final KnownRegistryDataMapsPayload payload, final ConfigurationPayloadContext context) {
+        record MandatoryEntry(ResourceKey<Registry<?>> registry, ResourceLocation id) {}
+        final Set<MandatoryEntry> ourMandatory = new HashSet<>();
+        getDataMaps().forEach((reg, values) -> values.values().forEach(attach -> {
+            if (attach.mandatorySync()) {
+                ourMandatory.add(new MandatoryEntry(reg, attach.id()));
+            }
+        }));
+
+        final Set<MandatoryEntry> theirMandatory = new HashSet<>();
+        payload.dataMaps().forEach((reg, values) -> values.forEach(attach -> {
+            if (attach.mandatory()) {
+                theirMandatory.add(new MandatoryEntry(reg, attach.id()));
+            }
+        }));
+
+        final var common = Sets.intersection(ourMandatory, theirMandatory);
+        if (common.size() != ourMandatory.size() || common.size() != theirMandatory.size()) {
+            final var missingOur = Sets.difference(common, ourMandatory);
+            final Set<MandatoryEntry> missing;
+            final String key;
+            if (!missingOur.isEmpty()) {
+                missing = missingOur;
+                key = "neoforge.network.data_maps.missing_our";
+            } else {
+                missing = Sets.difference(common, theirMandatory);
```

Comment:
`Sets.difference()` returns what is in the first set but absent from the second one, so order matters.
```suggestion
                missing = Sets.difference(theirMandatory, common);
```